File: src-tauri/src/models/resource.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub enum ResourceSource {
    Modrinth,
    CurseForge,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub enum ResourceType {
    Mod,
    Plugin,
    DataPack,
    Shader,
    ResourcePack,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ResourceSearchResult {
    pub id: String,
    pub name: String,
    pub slug: String,
    pub summary: String,
    pub author: String,
    pub downloads: u64,
    pub icon_url: Option<String>,
    pub source: ResourceSource,
    pub project_type: ResourceType,
    pub categories: Vec<String>,
    pub latest_game_versions: Vec<String>,
    pub date_modified: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ResourceVersion {
    pub id: String,
    pub version_number: String,
    pub name: String,
    pub game_versions: Vec<String>,
    pub loaders: Vec<String>,
    pub download_url: String,
    pub file_name: String,
    pub file_size: u64,
    pub dependencies: Vec<ResourceDependency>,
    pub date_published: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ResourceDependency {
    pub project_id: String,
    pub dependency_type: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ResourceDetail {
    pub id: String,
    pub name: String,
    pub description: String,
    pub source: ResourceSource,
    pub downloads: u64,
    pub icon_url: Option<String>,
    pub source_url: Option<String>,
    pub wiki_url: Option<String>,
    pub versions: Vec<ResourceVersion>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct InstalledResource {
    pub id: String,
    pub project_id: String,
    pub source: ResourceSource,
    pub name: String,
    pub version_id: String,
    pub version_number: String,
    pub file_name: String,
    pub installed_at: u64,
    pub server_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ResourceUpdateInfo {
    pub resource_id: String,
    pub project_id: String,
    pub current_version: String,
    pub latest_version_id: String,
    pub latest_version_number: String,
    pub source: ResourceSource,
    pub name: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SearchRequest {
    pub query: String,
    pub game_version: Option<String>,
    pub loader: Option<String>,
    pub project_type: Option<String>,
    pub source: Option<String>,
    pub offset: u32,
    pub limit: u32,
}

File: src/api/resource.ts
import { tauriInvoke } from "@api/tauri";

export interface ResourceSearchResult {
  id: string;
  name: string;
  slug: string;
  summary: string;
  author: string;
  downloads: number;
  icon_url: string | null;
  source: "Modrinth" | "CurseForge";
  project_type: "Mod" | "Plugin" | "DataPack" | "Shader" | "ResourcePack";
  categories: string[];
  latest_game_versions: string[];
  date_modified: string;
}

export interface ResourceVersion {
  id: string;
  version_number: string;
  name: string;
  game_versions: string[];
  loaders: string[];
  download_url: string;
  file_name: string;
  file_size: number;
  dependencies: ResourceDependency[];
  date_published: string;
}

export interface ResourceDependency {
  project_id: string;
  dependency_type: string;
}

export interface ResourceDetail {
  id: string;
  name: string;
  description: string;
  source: "Modrinth" | "CurseForge";
  downloads: number;
  icon_url: string | null;
  source_url: string | null;
  wiki_url: string | null;
  versions: ResourceVersion[];
}

export interface InstalledResource {
  id: string;
  project_id: string;
  source: "Modrinth" | "CurseForge";
  name: string;
  version_id: string;
  version_number: string;
  file_name: string;
  installed_at: number;
  server_id: string;
}

export interface ResourceUpdateInfo {
  resource_id: string;
  project_id: string;
  current_version: string;
  latest_version_id: string;
  latest_version_number: string;
  source: "Modrinth" | "CurseForge";
  name: string;
}

export const resourceApi = {
  async search(
    query: string,
    gameVersion: string | null,
    loader: string | null,
    projectType: string | null,
    source: string | null,
    offset: number,
    limit: number,
  ): Promise<ResourceSearchResult[]> {
    return tauriInvoke("search_resources", {
      query,
      gameVersion,
      loader,
      projectType,
      source,
      offset,
      limit,
    });
  },

  async getDetail(source: string, projectId: string): Promise<ResourceDetail> {
    return tauriInvoke("get_resource_detail", { source, projectId });
  },

  async getVersions(
    source: string,
    projectId: string,
    gameVersion: string | null,
    loader: string | null,
  ): Promise<ResourceVersion[]> {
    return tauriInvoke("get_resource_versions", {
      source,
      projectId,
      gameVersion,
      loader,
    });
  },

  async install(
    serverId: string,
    source: string,
    projectId: string,
    versionId: string,
  ): Promise<InstalledResource> {
    return tauriInvoke("install_resource", {
      serverId,
      source,
      projectId,
      versionId,
    });
  },

  async getInstalled(serverId: string): Promise<InstalledResource[]> {
    return tauriInvoke("get_installed_resources", { serverId });
  },

  async uninstall(serverId: string, resourceId: string): Promise<void> {
    return tauriInvoke("uninstall_resource", { serverId, resourceId });
  },

  async checkUpdates(serverId: string): Promise<ResourceUpdateInfo[]> {
    return tauriInvoke("check_resource_updates", { serverId });
  },

  async update(
    serverId: string,
    resourceId: string,
    versionId: string,
  ): Promise<InstalledResource> {
    return tauriInvoke("update_resource", { serverId, resourceId, versionId });
  },
};
